java/server/src/org/openqa/grid/web/servlet/NodeSessionsServlet.java

```diff
@@ -0,0 +1,154 @@
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * API to query all the sessions that are currently running in the hub.
+ */
+public class NodeSessionsServlet extends RegistryBasedServlet {
+
+  private final Json json = new Json();
+
+  public NodeSessionsServlet() {
+    this(null);
+  }
+
+  public NodeSessionsServlet(GridRegistry registry) {
+    super(registry);
+  }
+
+  @Override
+  protected void doGet(HttpServletRequest req, HttpServletResponse rsp) throws IOException {
+    process(rsp);
+  }
+
+  @Override
+  protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException {
+    if (req.getInputStream() != null) {
+      process(resp);
+    } else {
+      process(resp);
+    }
+  }
+
+  protected void process(HttpServletResponse response) throws IOException {
+    response.setContentType("application/json");
+    response.setCharacterEncoding("UTF-8");
+    response.setStatus(200);
+    Map<String, Object> proxies = new TreeMap<>();
+    try (Writer writer = response.getWriter(); JsonOutput out = json.newOutput(writer)) {
+      proxies.put("success", true);
+      proxies.put("proxies", extractSessionsFromAllProxies());
+      out.write(proxies);
+    }
+  }
+
+  private List<Map<String, Object>> extractSessionsFromAllProxies() {
+    List<Map<String, Object>> results = new LinkedList<>();
+    List<RemoteProxy> proxies = getRegistry().getAllProxies().getBusyProxies();
+    for (RemoteProxy proxy : proxies) {
+      Map<String, Object> sessionsInProxy = new TreeMap<>(extractSessionInfo(proxy));
+      if (sessionsInProxy.isEmpty()) {
+        continue;
+      }
+      Map<String, Object> res = new TreeMap<>();
+      res.put("proxyId", proxy.getId());
+      res.put("proxyRemoteHost", proxy.getRemoteHost().toString());
+      res.put("sessions", sessionsInProxy);
+      results.add(res);
+    }
+    return results;
+  }
+
+  private static Map<String, Object> extractSessionInfo(RemoteProxy proxy) {
+    try {
+      URL url = proxy.getRemoteHost();
+      HttpRequest req = new HttpRequest(HttpMethod.GET, "/wd/hub/sessions");
+      HttpResponse rsp = proxy.getHttpClient(url).execute(req);
+      try (InputStream in = new ByteArrayInputStream(asBytes(rsp));
+           Reader reader = new InputStreamReader(in, getContentEncoding(rsp))) {
+        Object body = new Json().newInput(reader).read(Object.class);
+        if (body instanceof Map) {
+          return (Map<String, Object>) body;
+        }
+      } catch (JsonException e) {
+        // Nothing to do --- poorly formed payload.
+      }
+    } catch (IOException e) {
+      throw new GridException(e.getMessage());
+    }
+    return new TreeMap<>();
+  }
+
+  private static byte[] asBytes(HttpResponse rsp) {
+    InputStream stream = rsp.consumeContentStream();
```

Comment:
Heavens. Just call `rsp.getContent()`. You're not saving any memory doing this.

Reply:
Fixed. I think I borrowed that code from somewhere :)